File: Server/src/instruments/instrumentService.ts
import { gunzipSync } from 'node:zlib'
import type { Pool } from 'pg'
import type { InstrumentRecord } from '../lib/contracts.js'
import { COMPANY_REFERENCE_DATA } from './companyReferenceData.js'

type UpstoxInstrumentRow = {
  trading_symbol?: string
  symbol?: string
  name?: string
  exchange?: string
  segment?: string
  instrument_type?: string
  instrument_key?: string
  instrumentKey?: string
}

const isInstrumentKeyConflict = (error: unknown) =>
  typeof error === 'object' &&
  error !== null &&
  'code' in error &&
  error.code === '23505' &&
  'constraint' in error &&
  error.constraint === 'instruments_instrument_key_key'

export class InstrumentService {
  constructor(private readonly pool: Pool) {}

  async syncFromUpstox(instrumentsUrl: string) {
    if (!instrumentsUrl) {
      return 0
    }

    const response = await fetch(instrumentsUrl)
    if (!response.ok) {
      throw new Error(`Instrument sync failed with HTTP ${response.status}`)
    }

    const payload = await this.parseUpstoxPayload(response, instrumentsUrl)
    const seenSymbols = new Set<string>()
    const seenInstrumentKeys = new Set<string>()
    let count = 0

    for (const item of payload) {
      const symbol = (item.trading_symbol ?? item.symbol ?? '').trim().toUpperCase()
      const companyName = (item.name ?? symbol).trim()
      const exchange = (item.exchange ?? 'NSE').trim().toUpperCase()
      const segment = (item.segment ?? '').trim().toUpperCase()
      const instrumentType = (item.instrument_type ?? '').trim().toUpperCase()
      const instrumentKey = (item.instrument_key ?? item.instrumentKey ?? '').trim()

      if (!symbol || !instrumentKey) {
        continue
      }

      if (segment && segment !== 'NSE_EQ') {
        continue
      }

      if (instrumentType && !['EQ', 'BE'].includes(instrumentType)) {
        continue
      }

      // Upstox can return duplicate equity rows; skip duplicates so startup stays idempotent.
      if (seenSymbols.has(symbol) || seenInstrumentKeys.has(instrumentKey)) {
        continue
      }

      seenSymbols.add(symbol)
      seenInstrumentKeys.add(instrumentKey)
      const referenceProfile = COMPANY_REFERENCE_DATA[symbol]
      const metadata = referenceProfile ? JSON.stringify({ companyOverview: referenceProfile }) : '{}'

      try {
        await this.pool.query(
          `
            insert into instruments (symbol, company_name, exchange, instrument_key, metadata, updated_at)
            values ($1, $2, $3, $4, $5::jsonb, now())
            on conflict (symbol)
            do update set
              company_name = excluded.company_name,
              exchange = excluded.exchange,
              instrument_key = excluded.instrument_key,
              metadata = case
                when excluded.metadata = '{}'::jsonb then instruments.metadata
                else coalesce(instruments.metadata, '{}'::jsonb) || excluded.metadata
              end,
              updated_at = now()
          `,
          [symbol, companyName, exchange, instrumentKey, metadata],
        )
        count += 1
      } catch (error) {
        if (isInstrumentKeyConflict(error)) {
          continue
        }

        throw error
      }
    }

    return count
  }

  async search(query: string, limit = 20): Promise<InstrumentRecord[]> {
    const trimmed = query.trim()
    if (!trimmed) {
      return []
    }

    const result = await this.pool.query<{
      symbol: string
      company_name: string
      exchange: string
      instrument_key: string
    }>(
      `
        select symbol, company_name, exchange, instrument_key
        from instruments
        where symbol ilike $1 or company_name ilike $1
        order by
          case when symbol ilike $2 then 0 else 1 end,
          symbol asc
        limit $3
      `,
      [`%${trimmed}%`, `${trimmed}%`, limit],
    )

    return result.rows.map((row) => ({
      symbol: row.symbol,
      companyName: row.company_name,
      exchange: row.exchange,
      instrumentKey: row.instrument_key,
    }))
  }

  async listSymbols() {
    const result = await this.pool.query<{ symbol: string }>('select symbol from instruments order by symbol asc')
    return result.rows.map((row) => row.symbol)
  }

  async getBySymbols(symbols: string[]) {
    if (!symbols.length) {
      return []
    }

    const result = await this.pool.query<{
      symbol: string
      company_name: string
      exchange: string
      instrument_key: string
    }>(
      `
        select symbol, company_name, exchange, instrument_key
        from instruments
        where symbol = any($1::text[])
      `,
      [symbols],
    )

    return result.rows.map((row) => ({
      symbol: row.symbol,
      companyName: row.company_name,
      exchange: row.exchange,
      instrumentKey: row.instrument_key,
    }))
  }
  private async parseUpstoxPayload(response: Response, instrumentsUrl: string) {
    const contentType = response.headers.get('content-type') ?? ''
    const shouldGunzip = instrumentsUrl.endsWith('.gz') || /gzip/i.test(contentType)

    if (!shouldGunzip) {
      return (await response.json()) as UpstoxInstrumentRow[]
    }

    const compressed = Buffer.from(await response.arrayBuffer())
    return JSON.parse(gunzipSync(compressed).toString('utf8')) as UpstoxInstrumentRow[]
  }
}

File: Server/app/services/instrument_service.py
from __future__ import annotations

import gzip
import json
from contextlib import asynccontextmanager
from typing import TYPE_CHECKING, Any, AsyncGenerator

import httpx
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.cache import cache_get, cache_set

if TYPE_CHECKING:
    from sqlalchemy.ext.asyncio import async_sessionmaker


class InstrumentService:
    def __init__(
        self,
        db: AsyncSession | None = None,
        *,
        session_factory: "async_sessionmaker[AsyncSession] | None" = None,
    ) -> None:
        self._db = db
        self._session_factory = session_factory

    @asynccontextmanager
    async def _session(self) -> AsyncGenerator[AsyncSession, None]:
        if self._db is not None:
            yield self._db
        elif self._session_factory is not None:
            async with self._session_factory() as session:
                yield session
        else:
            raise RuntimeError("InstrumentService requires a db session or session_factory")

    async def sync_from_upstox(self, instruments_url: str) -> int:
        if not instruments_url:
            return 0

        async with httpx.AsyncClient(timeout=60) as client:
            response = await client.get(instruments_url)
            response.raise_for_status()

            content_type = response.headers.get("content-type", "")
            if instruments_url.endswith(".gz") or "gzip" in content_type:
                payload = json.loads(gzip.decompress(response.content).decode("utf-8"))
            else:
                payload = response.json()

        seen_symbols: set[str] = set()
        seen_keys: set[str] = set()
        count = 0

        async with self._session() as db:
            for item in payload:
                symbol = (item.get("trading_symbol") or item.get("symbol") or "").strip().upper()
                company_name = (item.get("name") or symbol).strip()
                exchange = (item.get("exchange") or "NSE").strip().upper()
                segment = (item.get("segment") or "").strip().upper()
                instrument_type = (item.get("instrument_type") or "").strip().upper()
                instrument_key = (item.get("instrument_key") or item.get("instrumentKey") or "").strip()

                if not symbol or not instrument_key:
                    continue
                if segment and segment != "NSE_EQ":
                    continue
                if instrument_type and instrument_type not in ("EQ", "BE"):
                    continue
                if symbol in seen_symbols or instrument_key in seen_keys:
                    continue

                seen_symbols.add(symbol)
                seen_keys.add(instrument_key)

                from app.services.company_reference import COMPANY_REFERENCE_DATA
                from app.services.nifty50_fundamentals import NIFTY50_FUNDAMENTALS

                fundamentals = NIFTY50_FUNDAMENTALS.get(symbol)
                reference = COMPANY_REFERENCE_DATA.get(symbol)
                overview = None
                if fundamentals or reference:
                    overview = {
                        **(fundamentals or {}),
                        **(reference or {}),  # richer reference seed wins for the 8 detailed stocks
                    }
                metadata = json.dumps({"companyOverview": overview}) if overview else "{}"

                await db.execute(
                    text(
                        """
                        INSERT INTO instruments (symbol, company_name, exchange, instrument_key, metadata, updated_at)
                        VALUES (:symbol, :company_name, :exchange, :instrument_key, :metadata, now())
                        ON CONFLICT (symbol) DO UPDATE SET
                            company_name = EXCLUDED.company_name,
                            exchange = EXCLUDED.exchange,
                            instrument_key = EXCLUDED.instrument_key,
                            metadata = CASE
                                WHEN EXCLUDED.metadata = '{}' THEN instruments.metadata
                                ELSE (COALESCE(instruments.metadata, '{}')::jsonb || EXCLUDED.metadata::jsonb)::text
                            END,
                            updated_at = now()
                        """
                    ),
                    {
                        "symbol": symbol,
                        "company_name": company_name,
                        "exchange": exchange,
                        "instrument_key": instrument_key,
                        "metadata": metadata,
                    },
                )
                count += 1

            await db.commit()

        return count

    async def search(self, query: str, limit: int = 20) -> list[dict]:
        trimmed = query.strip()
        if not trimmed:
            return []

        async with self._session() as db:
            result = await db.execute(
                text(
                    """
                    SELECT symbol, company_name, exchange, instrument_key
                    FROM instruments
                    WHERE symbol ILIKE :pattern OR company_name ILIKE :pattern
                    ORDER BY CASE WHEN symbol ILIKE :exact THEN 0 ELSE 1 END, symbol ASC
                    LIMIT :limit
                    """
                ),
                {"pattern": f"%{trimmed}%", "exact": f"{trimmed}%", "limit": limit},
            )
            rows = result.all()

        return [
            {
                "symbol": r[0],
                "company_name": r[1],
                "exchange": r[2],
                "instrument_key": r[3],
            }
            for r in rows
        ]

    async def list_symbols(self) -> list[str]:
        async with self._session() as db:
            result = await db.execute(text("SELECT symbol FROM instruments ORDER BY symbol ASC"))
            return [r[0] for r in result.all()]

    async def get_by_symbols(self, symbols: list[str]) -> list[dict]:
        if not symbols:
            return []

        async with self._session() as db:
            result = await db.execute(
                text(
                    """
                    SELECT symbol, company_name, exchange, instrument_key
                    FROM instruments
                    WHERE symbol = ANY(:symbols)
                    """
                ),
                {"symbols": symbols},
            )
            return [
                {
                    "symbol": r[0],
                    "company_name": r[1],
                    "exchange": r[2],
                    "instrument_key": r[3],
                }
                for r in result.all()
            ]
